solutions/USACO dec2020-cowntagion.mdx

```diff
@@ -0,0 +1,94 @@
+---
+id: usaco-1062
+source: USACO Silver 2020 December
+title: Cowntagion
+author: Kevin Sheng
+---
+
+[Official Editorial](http://www.usaco.org/current/data/sol_prob1_silver_dec20.html)
+
+<LanguageSection>
+
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class Cowntagion {
+    public static void main(String[] args) throws IOException {
+        BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
+        int farmNum  = Integer.parseInt(read.readLine());
+        ArrayList<Integer>[] neighbors = new ArrayList[farmNum];
+        for (int f = 0; f < farmNum; f++) {
+            neighbors[f] = new ArrayList<>();
+        }
+        for (int i = 0; i < farmNum - 1; i++) {  // note: a connected graph with n nodes and n - 1 edges is always a tree
+            StringTokenizer path = new StringTokenizer(read.readLine());
+            int farm1 = Integer.parseInt(path.nextToken()) - 1;
+            int farm2 = Integer.parseInt(path.nextToken()) - 1;
+            neighbors[farm1].add(farm2);
+            neighbors[farm2].add(farm1);
+        }
+
+        int minDays = 0;
+        HashSet<Integer> visited = new HashSet<>(Collections.singletonList(0));
+        ArrayDeque<Integer> frontier = new ArrayDeque<>(Collections.singletonList(0));
+        // do a bfs of the tree
+        while (!frontier.isEmpty()) {
+            int current = frontier.poll();
+            // this stores all the farms that this farm should spread to for optimal spreading
+            ArrayList<Integer> spreadTo = new ArrayList<>();
+            for (int n : neighbors[current]) {
+                if (!visited.contains(n)) {
```

Comment:
why not have a boolean array?

Reply:
I mean they both do pretty much the same thing, no?

Reply:
~~unordered maps/hashsets piss me off~~
Boolean arrays are more common, faster, less memory, and more clear.

Reply:
I wouldn't say they're "clear", but the first 3 I agree.